src/HydraAuction/Hydra/Interface.hs

```diff
@@ -1,3 +1,74 @@
-module HydraAuction.Hydra.Interface (HydraEvent (..)) where
+{-# LANGUAGE StrictData #-}
 
-data HydraEvent = NodeCommitted | ContestationTimeEnded
+module HydraAuction.Hydra.Interface (
+  commandConstructorName,
+  getHydraEventKind,
+  HydraEventKind (..),
+  HydraEvent (..),
+  HydraCommand (..),
+) where
+
+-- Prelude imports
+import Prelude
+
+-- Haskell imports
+import Data.Text.Internal (Text)
+
+-- Cardano imports
+import Cardano.Api.UTxO qualified as UTxO
+
+-- Hydra imports
+
+import Hydra.Cardano.Api (Tx)
+import Hydra.Chain (HeadId)
+
+data HydraCommand
+  = Init
+  | GetUTxO
+  | NewTx Tx
+  | Commit UTxO.UTxO
+  | Close
+  | Fanout
+  deriving stock (Show)
+
+commandConstructorName :: HydraCommand -> Text
+commandConstructorName Init = "Init"
+commandConstructorName GetUTxO = "GetUTxO"
+commandConstructorName (NewTx _) = "NewTx"
+commandConstructorName (Commit _) = "Commit"
+commandConstructorName Close = "Close"
+commandConstructorName Fanout = "Fanout"
+
+-- FIXME: Add UTxO to HeadIsFinalized
+data HydraEvent
+  = GetUTxOResponse UTxO.UTxO
+  | TxSeen Tx
+  | Committed UTxO.UTxO
+  | HeadIsInitializing HeadId
+  | HeadIsOpen
+  | HeadIsClosed
+  | ReadyToFanout
+  | HeadIsFinalized
+  deriving stock (Eq)
+
+data HydraEventKind
+  = GetUTxOResponseKind
+  | TxSeenKind
+  | CommittedKind
+  | HeadIsInitializingKind
+  | HeadIsOpenKind
+  | HeadIsClosedKind
+  | ReadyToFanoutKind
+  | HeadIsFinalizedKind
+  deriving stock (Eq)
+
+getHydraEventKind :: HydraEvent -> HydraEventKind
```

Comment:
I think we can use makePrisms for `HydraEvent` to achieve something similar. We can also remove the reification of `AwaitedHydraEvent` and make it `type AwaitedHydraEvent = HydraEvent -> Bool` so the generated functions can be used to check which constructor is chosen

Reply:
> I think we can use makePrisms for HydraEvent to achieve something similar.

Thanks, that seems like exactly the question I was talking about in haskell-learn channel.

I would not change that now, cuz I am bad at prisms, but you could refactor that later is you want to.

> make it type AwaitedHydraEvent = HydraEvent -> Bool so the generated functions can be used to check which constructor is chosen

Functions do not have Eq and are more obscure to understand. Also ADT AwaitedHydraEvent is able have some other implementation than filtering, in theory.